File: 0935-knight-dialer/0935-knight-dialer.py
class Solution:
    def knightDialer(self, n: int) -> int:
        dp = {}
        MOD = 10**9 + 7
        dial = [
            [1, 2, 3], 
            [4, 5, 6],
            [7, 8, 9],
            ["*", 0, "#"],
        ]
        rows, cols = len(dial), len(dial[0])
        
        def dfs(i, j, cnt):
            if i < 0 or j < 0 or i >= rows or j >= cols or dial[i][j] in ["*", "#"]:
                return 0
            
            if cnt == n:
                return 1
            
            if (i, j, cnt) in dp:
                return dp[(i, j, cnt)]
            
            dp[(i, j, cnt)] = (((((((dfs(i + 2, j + 1, cnt + 1) +
                         dfs(i + 2, j - 1, cnt + 1)) % MOD +
                         dfs(i - 2, j + 1, cnt + 1)) % MOD +
                         dfs(i - 2, j - 1, cnt + 1)) % MOD +
                         dfs(i + 1, j + 2, cnt + 1)) % MOD +
                         dfs(i + 1, j - 2, cnt + 1)) % MOD +
                         dfs(i - 1, j + 2, cnt + 1)) % MOD +
                         dfs(i - 1, j - 2, cnt + 1)) % MOD
                    
            return dp[(i, j, cnt)]
        
        res = 0
        for r in range(rows):
            for c in range(cols):
                res = (res + dfs(r, c, 1)) % MOD

        return res

File: 0935-knight-dialer/0935-knight-dialer.cpp
class Solution {
public:
    int dp[5001][4][3] = {0}; 
    int mod = 1000000007; 
    int memo(int n, int i, int j){
        if(i<0 || j <0 || i>3 || j > 2 || (i ==3 && j == 0) || (i == 3 && j ==2)) return 0;
        if(n == 0) return 1; 
        if(dp[n][i][j] > 0) return dp[n][i][j]; 
        return dp[n][i][j] = (((((((
            memo(n-1, i+2, j-1)+ memo(n-1, i+2, j+1))%mod +memo(n-1, i-2, j-1))%mod+ 
            memo(n-1, i-2, j+1))%mod + memo(n-1, i+1, j-2))%mod+ memo(n-1, i+1, j+2))%mod+ 
            memo(n-1, i-1, j-2))%mod+  memo(n-1, i-1, j+2))%mod; 
    }
    int knightDialer(int n) {
        int res = 0; 
        for(int i=0; i<4; i++){
            for(int j =0; j<3; j++){
                res = (res+memo(n-1, i,j))%mod; 
            }
        }
        return res ;
    }
};